graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        location_to_predicates = _construct_location_to_predicate_list(match_query)
+
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = _collect_filters_to_first_location_in_match_traversal(
+                match_traversal, location_to_predicates)
+            new_match_traversals.append(new_match_traversal)
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
+    if isinstance(expression, BinaryComposition):
+        if isinstance(expression.left, ContextField):
+            context_field = expression.left
+        elif isinstance(expression.right, ContextField):
+            context_field = expression.right
+        elif expression.operator == u'||':
+            if expression.left == TrueLiteral or expression.right == TrueLiteral:
+                return TrueLiteral
+            else:
+                return expression
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in current_locations:
+            return TrueLiteral
+        else:
+            return expression
+    elif isinstance(expression, TernaryConditional):
+        if expression.predicate == TrueLiteral:
+            return expression.if_true
+        else:
+            return expression
+    elif isinstance(expression, BetweenClause):
+        lower_bound = expression.lower_bound
+        upper_bound = expression.upper_bound
+        if isinstance(lower_bound, ContextField) or isinstance(upper_bound, ContextField):
+            raise AssertionError(u'Found BetweenClause with ContextFields as lower/upper bounds. '
+                                 u'THis should never happen: {}'.format(expression))
+    else:
+        return expression
+
+
+def _construct_update_tags_visitor_fn(current_locations):
+    """Return an Expression updater using the given `current_locations`."""
+    def visitor_fn(expression):
+        return _update_tagged_expression(expression, current_locations)
+    return visitor_fn
+
+
+def _get_current_locations_from_match_traversals(match_traversals):
```

Comment:
The docstring is great -- why don't we adopt its naming everywhere and refer to `present_locations` rather than `current_locations` everywhere?

Bonus points if you write a full docstring and explain how come locations can be absent.